fw/http_parser.c

```diff
@@ -8600,45 +8710,116 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
 	__FSM_STATE(Resp_I_CC_n) {
 		TRY_STR_LAMBDA_fixup(&TFW_STR_STRING("no-cache"),
 			&parser->hdr, {
-			parser->_acc = TFW_HTTP_CC_NO_CACHE;
-		}, Resp_I_CC_n, Resp_I_Flag);
+			parser->cache_control.dir_flag = TFW_HTTP_CC_NO_CACHE;
+		}, Resp_I_CC_n, Resp_I_Flag_Maybe_Qal);
 		TRY_STR_LAMBDA_fixup(&TFW_STR_STRING("no-store"),
 			&parser->hdr, {
-			parser->_acc = TFW_HTTP_CC_NO_STORE;
+			parser->cache_control.dir_flag = TFW_HTTP_CC_NO_STORE;
 		}, Resp_I_CC_n, Resp_I_Flag);
 		TRY_STR_LAMBDA_fixup(&TFW_STR_STRING("no-transform"),
 			&parser->hdr, {
-			parser->_acc = TFW_HTTP_CC_NO_TRANSFORM;
+			parser->cache_control.dir_flag = TFW_HTTP_CC_NO_TRANSFORM;
 		}, Resp_I_CC_n, Resp_I_Flag);
 		TRY_STR_INIT();
 		__FSM_I_JMP(Resp_I_Ext);
 	}
 
 	__FSM_STATE(Resp_I_CC_p) {
 		TRY_STR_LAMBDA_fixup(&TFW_STR_STRING("public"), &parser->hdr, {
-			parser->_acc = TFW_HTTP_CC_PUBLIC;
+			parser->cache_control.dir_flag = TFW_HTTP_CC_PUBLIC;
 		}, Resp_I_CC_p, Resp_I_Flag);
 		TRY_STR_LAMBDA_fixup(&TFW_STR_STRING("private"), &parser->hdr, {
-			parser->_acc = TFW_HTTP_CC_PRIVATE;
-		}, Resp_I_CC_p, Resp_I_Flag);
+			parser->cache_control.dir_flag = TFW_HTTP_CC_PRIVATE;
+		}, Resp_I_CC_p, Resp_I_Flag_Maybe_Qal);
 		TRY_STR_LAMBDA_fixup(&TFW_STR_STRING("proxy-revalidate"),
 			&parser->hdr, {
-			parser->_acc = TFW_HTTP_CC_PROXY_REVAL;
+			parser->cache_control.dir_flag = TFW_HTTP_CC_PROXY_REVAL;
 		}, Resp_I_CC_p, Resp_I_Flag);
 		TRY_STR_INIT();
 		__FSM_I_JMP(Resp_I_Ext);
 	}
 
+	__FSM_STATE(Resp_I_Flag_Maybe_Qal) {
+		/*
+		 * no-cache or private with arguments as defined by rfc7234
+		 * sections 5.2.2.2 and 5.2.2.6
+		 */
+		parser->cache_control.filled = true;
+		if (c == '=') {
+			/* Qualified form cancels out the general flag */
+			if (parser->cache_control.dir_flag == TFW_HTTP_CC_PRIVATE)
+				resp->cache_ctl.flags |= TFW_HTTP_CC_PRIVATE_QUAL;
+			else
+				resp->cache_ctl.flags |= TFW_HTTP_CC_NO_CACHE_QUAL;
```

Comment:
Ok, I see the usage of `dir_flag` versus direct update in `cache_ctl.flags`. Otherwise we'll need at least one more state them. However, neither of the *_QUAL flags are actually used besides the tests, so they should be just removed (and the tests reworked to test the things in a different way).